bin/reth/src/db/mod.rs

```diff
@@ -237,4 +238,77 @@ impl<'a, DB: Database> DbTool<'a, DB> {
         std::fs::remove_dir_all(path).wrap_err("Dropping the database failed")?;
         Ok(())
     }
+
+    /// Duplicates table data within a range (`From<u64>`) into another database: `destination_db`.
+    pub(crate) fn dump_table_with_range<T: Table>(
+        &mut self,
+        from: u64,
+        to: u64,
+        destination_db: &mut Env<WriteMap>,
+    ) -> eyre::Result<()>
+    where
+        <T as reth_db::table::Table>::Key: From<u64>,
+    {
+        destination_db.update(|write_tx| {
+            let mut cursor_destination = write_tx.cursor_write::<T>()?;
+
+            self.db.view(|read_tx| {
+                let mut cursor_source = read_tx.cursor_read::<T>()?;
+
+                for row in cursor_source.walk(from.into())?.take((to - from) as usize) {
+                    let (key, value) = row?;
+                    cursor_destination.append(key, value)?;
+                }
+
+                Ok::<(), eyre::ErrReport>(())
+            })
+        })??
+    }
+
+    /// Duplicates table data into another database: `destination_db`. The reason for `TXMut` for
+    /// the source database is that we might want to call this method after making a dry-run
+    /// unwind.
+    pub(crate) fn dump_table<T: Table>(
+        source_db_tx: &mut <DB as DatabaseGAT<'_>>::TXMut,
+        destination_db: &Env<WriteMap>,
+    ) -> eyre::Result<()> {
+        destination_db.update(|write_tx| {
+            let mut cursor_source = source_db_tx.cursor_read::<T>()?;
+            let mut cursor_destination = write_tx.cursor_write::<T>()?;
+
+            let start_walker = cursor_source.first().transpose();
+            let walker = Walker::new(&mut cursor_source, start_walker);
+            for kv in walker {
+                let (k, v) = kv?;
+                cursor_destination.append(k, v)?;
+            }
+            Ok::<(), eyre::ErrReport>(())
+        })?
+    }
+
+    /// Duplicates table (DUPSORT) data into another database: `destination_db`. The reason for
+    /// `TXMut` for the source database is that we might want to call this method after making a
+    /// dry-run unwind.
+    pub(crate) fn dump_dupsort<T: DupSort>(
+        source_db_tx: &mut <DB as DatabaseGAT<'_>>::TXMut,
+        destination_db: &Env<WriteMap>,
+    ) -> eyre::Result<()>
+    where
+        <T as DupSort>::SubKey: Default,
+    {
+        destination_db.update(|write_tx| {
+            let mut cursor_source = source_db_tx.cursor_dup_read::<T>()?;
+            let mut cursor_destination = write_tx.cursor_dup_write::<T>()?;
+
+            if let Some((first_key, _)) = cursor_source.first()? {
+                let walker = cursor_source.walk_dup(first_key, T::SubKey::default())?;
+                for kv in walker {
+                    let (k, v) = kv?;
+                    cursor_destination.append_dup(k, v)?;
+                }
+            }
+
+            Ok::<(), eyre::ErrReport>(())
+        })?
+    }
 }
```

Comment:
Let's move all these to the `reth_providers` crate and implement it for all `DbTx` types.

I think the main follow-up I care about here is creating an actual database and dumping it to disk and allow us to re-use it in an execution stage benchmark, like the other tests in stages/benches/criterion.rs.

I would try it out, see if you can get a small enough database that we can host somewhere in AWS/S3.

If that does not work, let's figure out all the touched storage slots (maybe by instrumenting the execution stage) so that we can actually trim down the state.